trunk/engine/mapedit/tilestack.3
.\"	$OpenBSD$
.\"
.\" Copyright (c) 2002 CubeSoft Communications, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistribution of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Neither the name of CubeSoft Communications, nor the names of its
.\"    contributors may be used to endorse or promote products derived from
.\"    this software without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd August 20, 2002
.Dt OBJQ 3
.Os
.Sh NAME
.Nm tilestack
.Nd agar tile stack widget
.Sh SYNOPSIS
.Bd -literal
#include <engine/mapedit/mapedit.h>
#include <engine/mapedit/tilestack.h>
.Ed
.Sh DESCRIPTION
The
.Nm
widget displays a list of objects available for edition.
.Sh INITIALIZATION
.nr nS 1
.Ft struct tilestack *
.Fn tilestack_new "struct region *reg" "int flags" "int rw" "int rh" "struct mapview *mv"
.Pp
.Ft void
.Fn tilestack_init "struct tilestack *ts" "int flags" "int rw" "int rh" "struct mapview *mv"
.nr nS 0
.Pp
The
.Fn tilestack_new
function allocates, initializes, and attaches a
.Nm
widget.
The
.Fn tilestack_init
function initializes a
.Nm
widget of size
.Fa rw
x
.Fa rh .
The
.Fa flags
cannot be changed at runtime, and may include:
.Pp
.Bd -literal
#define TILESTACK_HORIZ		0x01	/* Display tiles horizontally */
#define TILESTACK_VERT		0x02	/* Display tiles vertically */
.Ed
.Sh EVENTS
The
.Nm
widget reacts to the following events:
.Pp
.Bl -tag -width 25n -compact
.It widget-scaled
Adjust the tile display.
.El
.Pp
The
.Nm
widget does not generate any event.
.Sh SEE ALSO
.Xr agar 3 ,
.Xr mapview 3 ,
.Xr tilestack 3
.Sh HISTORY
The
.Nm
widget first appeared in AGAR 1.0.


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.8.0] - Unreleased
### Added
- [**Style Editor**](https://libagar.org/man3/AG_StyleEditor): Display computed style rules in addition to per-instance overrides.
- [**AG_Button**](https://libagar.org/man3/AG_Button): New style attributes `text-align` and `vertical-align`.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Introduce the `style()` method, called by the style compiler to parse widget-specific style attributes into an efficient binary representation.
- [**AG_Object**](https://libagar.org/man3/AG_Object): New function `AG_AccessVariable_NoLock()` (lockless variant of `AG_AccessVariable()`).
- [**AG_Object**](https://libagar.org/man3/AG_Object): Document the `AG_OBJECT_BOUND_EVENTS` flag.

### Removed
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove the "window-detached" event. It is redundant since the generic "detached" event does the same job.

### Changed
- Major performance improvements in the style compiler. Since it runs whenever widgets are resized, this should lead to better fluidity and overall lower CPU usage when Agar windows and container widgets are resized.
- [**AG_Object**](https://libagar.org/man3/AG_Object): In `AG_ObjectDestroy()`, do not call `AG_ObjectDetach()` on child objects. Instead, just cancel any active timers and null their parent pointers without raising a "detached" event. This allows us to perform garbage collection more efficiently.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Make `AG_TLIST_NO_BGLINES` the default.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): In `AG_TlistNewPolledMs()`, accept a `ms` argument of 0 to disable periodical updates (the caller may still trigger updates with "tlist-poll").
- In `AG_WidgetCopyStyle()`, prevent redundant locking by using `AG_AccessVariable_NoLock()`.

### Fixed
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Fix redundant "detached" handler invocations. Improve overall efficiency of the garbage-collection step after windows are destroyed.
- [**AG_Console**](https://libagar.org/man3/AG_Console): Prevent a redundant lock operation in the backend of `AG_ConsoleMsg()`.

## [1.7.1] - 2025-11-15
### Added
- An awesome new Agar logo. Thanks [eightbit](https://8bitprodigy.github.io)!
- Added support for the [Cmake](https://cmake.org) build system.
- Added support for the [Meson](https://mesonbuild.com) build system. Thanks [coleman](https://git.sr.ht/~coleman/)!
- New functions [**AG_InitVideoSDL2**](https://libagar.org/man3/AG_InitVideoSDL2) and [**AG_SetVideoSurfaceSDL2**](https://libagar.org/man3/AG_SetVideoSurfaceSDL2) for integrating with an existing SDL2 display context. Thanks Brigham Keys!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Add "Any File" type.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New function `AG_TlistCopy()`. Copy all items from a source to a destination `AG_Tlist`.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New constructor `AG_ComboNewFn()`. Sets a `combo-expanded` handler.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New member `nVisItems`. Set the number of items to show by default in expansions.
- [**AG_Checkbox**](https://libagar.org/man3/AG_Checkbox): New functions `AG_CheckboxText()` and `AG_CheckboxTextS()` to update the text label.
- Tweak the default color scheme. Document style defaults as a comment block in `gui/style.css`.
- New Ada bindings for the `Window`, `Box`, `Button`, `Checkbox` and `Combo` widget classes. Updated `ada/gui/demo`.

### Removed
- [**AG_Db**](https://libagar.org/man3/AG_Db): Remove support for DB5.
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove an unnecessary counter variable in `SizeAllocate()`.

### Changed
- Small cosmetic changes to the default color scheme. Set default font size back to 14pt.
- [**AG_Titlebar**](https://libagar.org/man3/AG_Titlebar): Enable the `AG_BUTTON_EXCL` optimization on titlebar buttons.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Allow the dropdown list to be populated by a `combo-expanded` handler.
- [**AG_UCombo**](https://libagar.org/man3/AG_UCombo): Allow the dropdown list to be populated by a `ucombo-expanded` handler.

### Fixed
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Make it again possible to statically initialize `list` before `combo-expanded`. Restores compatibility pre-1.6. Thanks Wally!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix widget geometries not updating when switching to a different Type filter.
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix compilation where `GLOB_TILDE` is unavailable.
- SDL2 drivers: Require at least version 2.0.22 of SDL2 (`SDL_HINT_MOUSE_AUTO_CAPTURE` is necessary for best user experience).
- [**AG_ProgressBar**](https://libagar.org/man3/AG_ProgressBar): Make `padding` work as expected in progress bar. Thanks scaramacai!
- [**cocoa**](https://libagar.org/man3/AG_DriverCocoa): Fix a compilation error.
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Fix a keyboard handling error. Don't pass `KeyRelease` events to `XwcLookupString()`. Fixes issue #74. Thanks @tsesani and @scaramacai!
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Introduce the `no-im` (disable input methods) driver option.
- [**MAP**](https://libagar.org/man3/MAP): Fix a crash when attempting to create an object if `pLibsRoot` is NULL.
- [**AG_Web**](https://libagar.org/man3/AG_Web): Fix inline compilation problem due to a typo in `net/web.h`.
- On Darwin platforms, build with `_DARWIN_C_SOURCE` globally. Fixes issue #77. Thanks @kwhr0!
- On FreeBSD check for availability of `CLOCK_SECOND` in addition to `HAVE_CLOCK_GETTIME`.
- Test for the `dirfd()` function at compilation.

engine/mapedit/tilestack.c
/*	$Csoft: tilestack.c,v 1.17 2002/12/24 10:30:53 vedge Exp $	*/

/*
 * Copyright (c) 2002 CubeSoft Communications, Inc. <http://www.csoft.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistribution of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Neither the name of CubeSoft Communications, nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <engine/engine.h>

#include <engine/map.h>
#include <engine/view.h>

#include <engine/widget/widget.h>
#include <engine/widget/window.h>
#include <engine/widget/primitive.h>
#include <engine/widget/text.h>

#include "tilestack.h"
#include "mapview.h"
#include "mapedit.h"

static const struct widget_ops tilestack_ops = {
	{
		widget_destroy,
		NULL,	/* load */
		NULL	/* save */
	},
	tilestack_draw,
	NULL		/* update */
};

enum {
	SELECTION_COLOR,
	GRID_COLOR
};

static void	tilestack_scaled(int, union evarg *);

struct tilestack *
tilestack_new(struct region *reg, int flags, int rw, int rh, struct mapview *mv)
{
	struct tilestack *tilestack;

	tilestack = emalloc(sizeof(struct tilestack));
	tilestack_init(tilestack, flags, rw, rh, mv);

	region_attach(reg, tilestack);

	return (tilestack);
}

void
tilestack_init(struct tilestack *ts, int flags, int rw, int rh,
    struct mapview *mv)
{
	widget_init(&ts->wid, "tilestack", &tilestack_ops, rw, rh);
	widget_map_color(ts, SELECTION_COLOR, "tilestack-selection", 0, 200, 0);
	widget_map_color(ts, GRID_COLOR, "tilestack-grid", 208, 208, 208);

	ts->offs = 0;
	ts->flags = (flags != 0) ? flags : TILESTACK_VERT;
	ts->mv = mv;

	event_new(ts, "widget-scaled", tilestack_scaled, NULL);
}

static void
tilestack_scaled(int argc, union evarg *argv)
{
	struct tilestack *ts = argv[0].p;
	int maxw = argv[1].i, maxh = argv[2].i;
	int w, h;

	for (w = 0; w < WIDGET(ts)->w-1 && w < maxw; w += TILEW) ;;
	for (h = 0; h < WIDGET(ts)->h-1 && h < maxh; h += TILEH) ;;
	
	WIDGET(ts)->w = w;
	WIDGET(ts)->h = h;
}

void
tilestack_draw(void *p)
{
	struct tilestack *ts = p;
	struct mapview *mv = ts->mv;
	struct node *n;
	struct noderef *nref;
	struct art_anim *an;
	int nx, ny, y = 0;
	SDL_Surface *su;

	nx = mv->mx + mv->mouse.x;
	ny = mv->my + mv->mouse.y;

	if (nx >= mv->map->mapw - 1 || ny >= mv->map->maph ||
	    nx < 0 || ny < 0) {
		return;
	}

	n = &mv->map->map[ny][nx];

	TAILQ_FOREACH(nref, &n->nrefs, nrefs) {
		switch (nref->type) {
		case NODEREF_SPRITE:
			su = view_scale_surface(SPRITE(nref->pobj, nref->offs),
			    WIDGET(ts)->w, WIDGET(ts)->w);
			widget_blit(ts, su, 0, y);
			SDL_FreeSurface(su);
			break;
		case NODEREF_ANIM:
			an = ANIM(nref->pobj, nref->offs);
			su = view_scale_surface(an->frames[0],
			    WIDGET(ts)->w, WIDGET(ts)->w);
			widget_blit(ts, su, 0, y);
			SDL_FreeSurface(su);

			if (nref->data.anim.flags & NODEREF_ANIM_AUTO) {
				widget_blit(ts,
				    SPRITE(mv->med, MAPEDIT_ANIM_DELTA_TXT),
				    0, y);
			}
			widget_blit(ts, SPRITE(mv->med, MAPEDIT_ANIM_TXT),
			    0, y);
			break;
		case NODEREF_WARP:
			su = text_render(NULL, -1,
			    SDL_MapRGB(view->v->format, 255, 255, 255),
			    nref->data.warp.map);
			widget_blit(ts, su, 0, y);
			SDL_FreeSurface(su);
			break;
		}

		primitives.rect_outlined(ts, 0, y, TILEW, TILEH,
		    WIDGET_COLOR(ts, GRID_COLOR));

		y += TILEH;
		if (y + TILEH > WIDGET(ts)->h) {
			break;
		}
	}
}

